processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -271,19 +273,52 @@ public File merge(
     );
   }
 
+  private List<String> getLongestSharedDimOrder(List<IndexableAdapter> indexes)
+  {
+    int maxSize = 0;
+    Iterable<String> orderingCandidate = null;
+    for (IndexableAdapter index : indexes) {
+      int iterSize = Iterators.size(index.getDimensionNames().iterator());
+      if (iterSize > maxSize) {
+        maxSize = iterSize;
+        orderingCandidate = index.getDimensionNames();
+      }
+    }
+
+    if (orderingCandidate == null) {
+      return null;
+    }
+
+    for (IndexableAdapter index : indexes) {
+      Iterator<String> candidateIter = orderingCandidate.iterator();
+      for (String matchDim : index.getDimensionNames()) {
+        boolean matched = false;
+        while (candidateIter.hasNext()) {
```

Comment:
@himanshug [A,B,C,D] ordering would be the same as lexicographic ordering

If there is no ordering that covers all sub orderings, then we have to generate an ordering that includes all dimensions, and lexicographic seemed like a sensible fallback choice.